test/unit/StakingPool/calculatePremium.js

```diff
@@ -0,0 +1,611 @@
+      expect(product.nextPriceUpdateTime).to.be.equal(timestamp);
+    }
+  });
+
+  it('should return 0 premium when period is 0', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { NXM_PER_ALLOCATION_UNIT } = this.config;
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = 0;
+    const coverAmountRaw = parseEther('100');
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+
+    const totalCapacity = parseEther('100');
+    const initialCapacityUsed = BigNumber.from(0);
+
+    const { premium } = await stakingPool.calculatePremium(
+      stakedProduct,
+      period,
+      coverAmount,
+      initialCapacityUsed,
+      totalCapacity,
+      stakedProduct.targetPrice,
+      timestamp,
+    );
+    expect(premium).to.be.equal(0);
+  });
+
+  it('should overflow when calculating premium for very large capacity', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { NXM_PER_ALLOCATION_UNIT } = this.config;
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = daysToSeconds(365);
+    const coverAmountRaw = parseEther('100');
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+
+    const totalCapacity = BigNumber.from(2).pow(255);
+    const initialCapacityUsed = BigNumber.from(0);
+    const basePrice = calculateBasePrice(timestamp, stakedProduct, this.config.PRICE_CHANGE_PER_DAY);
+
+    await expect(
+      stakingPool.calculatePremium(
+        stakedProduct,
+        period,
+        coverAmount,
+        initialCapacityUsed,
+        totalCapacity,
+        stakedProduct.targetPrice,
+        timestamp,
+      ),
+    ).to.be.revertedWithPanic(UNDER_OR_OVERFLOW);
+    await expect(
+      stakingPool.calculatePremiumPerYear(basePrice, coverAmount, initialCapacityUsed, totalCapacity),
+    ).to.be.revertedWithPanic(UNDER_OR_OVERFLOW);
+  });
+
+  it('should calculate the premium correctly when cover amount is equal to total capacity', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { INITIAL_PRICE_DENOMINATOR, NXM_PER_ALLOCATION_UNIT } = this.config;
+    // call staking pool and calculate premium
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = daysToSeconds(365);
+    const coverAmountRaw = BigNumber.from(2).pow(96);
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+    const initialCapacityUsed = 0;
+    const totalCapacity = coverAmount;
+    const basePrice = calculateBasePrice(timestamp, stakedProduct, this.config.PRICE_CHANGE_PER_DAY);
+    const { premium: expectedPremium } = await stakingPool.calculatePremium(
```

Comment:
the naming here is the other way around

expectedPremium should be a value you compute in your test code;

what comes out of `stakingPool.calculatePremium` is the `actualPremium` - what is being tested.

`actualPremium` what the function under test (solidity) says the result is
`expectedPremium` what the test code expects the function being tested to return